Replace tectonic-network-operator with cluster-network-operator

squeed opened the pull request:
This removes tectonic-network-operator and replaces it with the cluster-network-operator.

This won't work quite yet, because there is a bug in origin. But how we generate the configuration is up for review.

Network configuration is a funny beast. We basically can't change it once the cluster is up. However, there is a distinct "happy path" along with a large number of possible settings tweaks. So, right now this generates a sane network configuration from the installer network configuration.

We probably want to figure out a better way to do that at some point; the installer config doesn't capture even the full set of IP address options.

abhinavdahiya commented on pkg/asset/manifests/network-operator.go:
```diff
@@ -1,72 +1,117 @@
+				},
+			},
+			DefaultNetwork: netopv1.DefaultNetworkDefinition{
+				Type: netopv1.NetworkTypeOpenshiftSDN,
```
from install config `Networking.Type`

wking commented on pkg/asset/manifests/network-operator.go:
```diff
@@ -1,72 +1,117 @@
+	// we need to manually create our CRD first, so we can create the
+	// configuration instance of it
+	netConfigCRD = `
+---
```
Most of our YAML templates don't use an explicit document-start `---`:

```console
$ git grep apiVersion pkg/asset | wc -l
43
$ git grep -- --- pkg/asset | wc -l
16
```

Can we drop this line to try and consolidate around that approach?

wking commented on pkg/asset/manifests/network-operator.go:
```diff
@@ -1,72 +1,117 @@
 }
 
-// Generate generates the network-operator-config.yml and network-operator-manifest.yml files
+// Generate generates the network operator config and its CRD
```
nit: This was true of the previous comment as well, but can we get a trailing period here?  ([docs][1])

[1]: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

squeed commented:
Okay, refactored a bit. This change is now somewhat broader in scope.

I deprecated the badly-named `podCIDR` variable in the install-config, and added a correct multiple-cluster-cidr variable. We do need to think if we want to expose the complete network-operator API directly in the install-config, or if we want to just support the happy-path and leave the rest to customization.

Another open question is what the user should do when they change their cidrs. We actually will support expanding the service and cluster ip ranges. These changes will have to be propagated to their various consumers.

abhinavdahiya commented on pkg/asset/manifests/cluster_k8s_io.go:
```diff
@@ -0,0 +1,62 @@
+package manifests
```
already have this file here `pkg/assets/machines/...`

squeed replied:
import loop. had to move it.

abhinavdahiya replied:
split commits so that's visible. and include description of how there is an import loop. I want to keep it in the same package as much as possible.

squeed replied:
hm, guess enough of the file changed that it didn't trigger the git rename heuristic. neat.

abhinavdahiya commented:
> I deprecated the badly-named podCIDR variable in the install-config, and added a correct multiple-cluster-cidr variable. We do need to think if we want to expose the complete network-operator API directly in the install-config, or if we want to just support the happy-path and leave the rest to customization.

InstallConfig is supposed to provide very simple options. Why expose `[]ClusterCIDR` ie list of these and if the NetworkOperator supports expanding these in running cluster allow users to give only one and customize later on.

squeed commented:
> InstallConfig is supposed to provide very simple options. Why expose `[]ClusterCIDR` ie list of these and if the NetworkOperator supports expanding these in running cluster allow users to give only one and customize later on.

Oh, it's so much muddier than that.

First of all, the network-operator currently doesn't support rolling out any changes, and it probably won't for a while. So we can't support that. Also, we won't ever support changes for non-Openshift-maintained network plugins. I'd also just like to avoid network changes as much as possible. They're always risky, even when done correctly.

Right now, the address configurations are read in to the installer, then split out in to the `install-config`, `NetworkConfig.networkoperator.openshift.io`, and  `Cluster.cluster.k8s.io`. Too many things, right now, read from `install-config`, including the kube-apiserver-operator. Right now things are fine because IPs are immutable.

In the future, we stop reading from `install-config` and just use `Cluster.cluster.k8s.io` for addressing information -  and manage that that under network-operator reconciliation. The `Cluster` object isn't exclusively the domain of the `network-operator`, though, so we need to be somewhat subtle.

Even when we reach that ideal state, it's not clear how we should plumb through IP space configuration in the installer. Since IP blocks are general-case immutable, users do need to choose them correctly (or else!), so we need it to be part of even the happy-path configuration. However, any subsequent changes will presumably be made to the `NetworkConfig` object directly, leaving the `install-config` object in cluster out of date. Is that OK? Is there a way to give the installer IP configuration without writing it to the cluster's `install-config`?

squeed commented:
Filed openshift/cluster-kube-apiserver-operator#106 to stop reading cidrs from the install-config.

squeed commented:
I had to refactor this slightly one more time: the network-operator was setting `KUBERNETES_SVC_HOST` to 127.0.0.1, but that doesn't work because the bootstrap apiserver is still active. So I now have the installer creating an additional config map.

The alternative would be to special-casing the network-operator and running it on the bootstrap node. That seems uglier.

squeed commented:
A few more things to work out:
- Can the network-operator run in runlevel 1? I don't think so, but would love to be wrong
- If not, is there a better way to pass the apiserver URL?

knobunc commented:
@squeed if you want this to go in you need to remove [wip] from the title.

enxebre commented on pkg/asset/manifests/cluster_k8s_io.go:
```diff
@@ -0,0 +1,65 @@
+		return errors.Wrapf(err, "Could not generate ClusterNetworkingConfig")
+	}
+
+	cluster := clusterv1a1.Cluster{
```
Also I'm thinking this might be relevant for a 3.11 to 4.0 upgrade

squeed commented on pkg/asset/manifests/network.go:
```diff
@@ -0,0 +1,252 @@
+	if err != nil { // impossible
+		return errors.Wrapf(err, "invalid apiserver url")
+	}
+	no.configMap = &corev1.ConfigMap{
```
yeah, i think the right way is to bind-mount in the host's kubeconfig and merge in the serviceaccount token. solves a lot of problems.

squeed replied:
Done.

squeed commented:
> if operators should use `Cluster.cluster.k8s.io` object for cidrs, the network operator shuold do that same right? so the network operator config should drop that part?

@abhinavdahiya The other way around. The network config has more detail than the cluster config (it has podcidr length in addition to full cluster cidrs). Then the operator should update the cluster config.

squeed commented:
OK, removed that (awful) ConfigMap; now the operator is configured via a kubeconfig bind-mount.

This should be ready to go - once openshift/cluster-network-operator#28 merges and is picked up by the release, this should pass tests. It Works On My Machine(tm).

abhinavdahiya commented:
> > if operators should use `Cluster.cluster.k8s.io` object for cidrs, the network operator shuold do that same right? so the network operator config should drop that part?
> 
> @abhinavdahiya The other way around. The network config has more detail than the cluster config (it has podcidr length in addition to full cluster cidrs). Then the operator should update the cluster config.

@sttts @smarterclayton @squeed 
Do you think this is best chain of ownership/consumption:
the network operator is going to reconcile the options from its networkconfig to `Cluster.cluster.k8s.io` and all other operators that need access to cluster,service cidrs get the info from `Cluster.cluster.k8s.io`

* A user when trying to update ClusterCIDR, updates the networlconfig and network operator reconcile the `Cluster.cluster.k8s.io` to match the new cidrs and that trickles down to other operators....
* `Cluster.cluster.k8s.io` name is usually clustername, so discovery for operators means list and use the first one.

smarterclayton commented:
So not sure what cluster.cluster.k8s.io is, but the "network config" for
the cluster will probably be in an object called "
network.config.openshift.io".  The master team was working on splitting
these out, but the control plane operators would be expecting to read from
that object into their own config.

I would expect the installer to translate to network.config.openshift.io
directly and everyone to read from it.  @sttts and @deads2k have been the
most involved from it.

smarterclayton commented:
Here's how "Image" and "Build" global config look:

https://github.com/openshift/api/blob/master/config/v1/types.go#L14

David has been driving this but is neck deep.  The idea is that there would
be "Network" in this same group and Network would have network settings.
Installer would directly write Network from its config.

On Wed, Nov 7, 2018 at 3:51 PM Clayton Coleman <ccoleman@redhat.com> wrote:

> So not sure what cluster.cluster.k8s.io is, but the "network config" for
> the cluster will probably be in an object called "
> network.config.openshift.io".  The master team was working on splitting
> these out, but the control plane operators would be expecting to read from
> that object into their own config.
>
> I would expect the installer to translate to network.config.openshift.io
> directly and everyone to read from it.  @sttts and @deads2k have been the
> most involved from it.
>

squeed commented:
> . I would expect the installer to translate to network.config.openshift.io directly and everyone to read from it. @sttts and @deads2k have been the most involved from it.

This PR has the installer translate to the network operator CRD. Do we really need an additional object? Should we kill the network operator one?

squeed commented:
/retest

squeed commented:
In any case, deciding exactly how the configuration is plumbed shouldn't, IMHO, block this PR. It's important that we kill the tectonic-network-operator.

squeed commented:
/retest

knobunc commented:
@smarterclayton I completely understand you concerns but agree with @squeed.  I would really like to get this integrated into the installer and then fix up the issues you've identified.  We need to start getting soak time on this in the installer ASAP.

squeed commented:
/retest

smarterclayton commented:
Yes, the teams priority is getting integrated.  We will do the refactoring
post merge but before we lock the api down to ship to dedicated.  Plan for
it to happen.

On Nov 8, 2018, at 10:23 AM, Ben Bennett <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I completely understand
you concerns but agree with @squeed <https://github.com/squeed>. I would
really like to get this integrated into the installer and then fix up the
issues you've identified. We need to start getting soak time on this in the
installer ASAP.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/installer/pull/600#issuecomment-437034159>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p2Kr8zbHxrsWynjnh25D56XBNsmRks5utExcgaJpZM4YMLXc>
.

abhinavdahiya commented:
so testing this locally on livirt:
```console
$ oc get nodes,pods --all-namespaces
NAME                            STATUS    ROLES     AGE       VERSION
node/adahiya-0-master-0         Ready     master    47m       v1.11.0+d4cacc0
node/adahiya-0-worker-0-22h5z   Ready     worker    42m       v1.11.0+d4cacc0

NAMESPACE                                NAME                                                                  READY     STATUS             RESTARTS   AGE
kube-system                              pod/etcd-member-adahiya-0-master-0                                    1/1       Running            0          47m
kube-system                              pod/kube-scheduler-tdrf6                                              1/1       Running            0          47m
kube-system                              pod/metrics-server-5767bfc576-bjxx8                                   2/2       Running            0          42m
openshift-apiserver                      pod/apiserver-ctrsc                                                   1/1       Running            2          43m
openshift-cluster-api                    pod/clusterapi-manager-controllers-78686c5666-slj98                   3/3       Running            0          44m
openshift-cluster-api                    pod/machine-api-operator-5b4b8ccf58-z8bkl                             1/1       Running            0          47m
openshift-cluster-dns-operator           pod/cluster-dns-operator-6dc9574c85-vgdg9                             1/1       Running            0          47m
openshift-cluster-dns                    pod/dns-default-flrkc                                                 1/1       Running            0          44m
openshift-cluster-dns                    pod/dns-default-t7gxh                                                 1/1       Running            0          42m
openshift-cluster-network-operator       pod/cluster-network-operator-vlr59                                    1/1       Running            0          47m
openshift-cluster-node-tuning-operator   pod/cluster-node-tuning-operator-cb6c47fd5-nwk4h                      1/1       Running            0          42m
openshift-cluster-node-tuning            pod/tuned-27txf                                                       1/1       Running            0          38m
openshift-cluster-node-tuning            pod/tuned-xvnl5                                                       1/1       Running            0          38m
openshift-cluster-samples-operator       pod/cluster-samples-operator-7d44b86d67-dxmch                         1/1       Running            0          42m
openshift-cluster-version                pod/cluster-version-operator-9bktg                                    1/1       Running            6          47m
openshift-console                        pod/console-operator-9fb59b9-5wdh6                                    1/1       Running            0          42m
openshift-console                        pod/openshift-console-7f897ffd5c-gcvh9                                1/1       Running            0          38m
openshift-controller-manager             pod/controller-manager-x9rfz                                          1/1       Running            0          44m
openshift-core-operators                 pod/openshift-cluster-kube-apiserver-operator-84f9f4f457-ffk9w        1/1       Running            0          47m
openshift-core-operators                 pod/openshift-cluster-kube-controller-manager-operator-78c54b8p8xf4   1/1       Running            0          47m
openshift-core-operators                 pod/openshift-cluster-kube-scheduler-operator-8675bf756b-jrjz7        1/1       Running            0          47m
openshift-core-operators                 pod/openshift-cluster-openshift-apiserver-operator-67cbc799b9-x67kk   1/1       Running            1          47m
openshift-core-operators                 pod/openshift-cluster-openshift-controller-manager-operator-98svk45   1/1       Running            0          47m
openshift-core-operators                 pod/openshift-service-cert-signer-operator-74d7fc9f45-df5n9           1/1       Running            0          47m
openshift-csi-operator                   pod/csi-operator-677bcb6bf-hvj24                                      1/1       Running            0          42m
openshift-image-registry                 pod/cluster-image-registry-operator-74f74c94b8-gd82j                  1/1       Running            0          42m
openshift-ingress-operator               pod/ingress-operator-7f594966d6-k9xsw                                 1/1       Running            0          42m
openshift-ingress                        pod/router-default-k8klh                                              1/1       Running            0          39m
openshift-kube-apiserver                 pod/installer-1-adahiya-0-master-0                                    0/1       Completed          0          44m
openshift-kube-apiserver                 pod/openshift-kube-apiserver-adahiya-0-master-0                       1/1       Running            0          44m
openshift-kube-controller-manager        pod/installer-1-adahiya-0-master-0                                    0/1       Completed          0          44m
openshift-kube-controller-manager        pod/openshift-kube-controller-manager-adahiya-0-master-0              1/1       Running            0          44m
openshift-kube-scheduler                 pod/installer-1-adahiya-0-master-0                                    0/1       Completed          0          45m
openshift-kube-scheduler                 pod/openshift-kube-scheduler-adahiya-0-master-0                       1/1       Running            0          44m
openshift-machine-config-operator        pod/machine-config-controller-6986f78f6f-fjjw7                        1/1       Running            0          45m
openshift-machine-config-operator        pod/machine-config-daemon-hvp7m                                       1/1       Running            2          42m
openshift-machine-config-operator        pod/machine-config-daemon-svpck                                       1/1       Running            0          44m
openshift-machine-config-operator        pod/machine-config-operator-558696bfb7-szzx5                          1/1       Running            0          47m
openshift-machine-config-operator        pod/machine-config-server-tpnw8                                       1/1       Running            0          44m
openshift-monitoring                     pod/alertmanager-main-0                                               3/3       Running            0          34m
openshift-monitoring                     pod/alertmanager-main-1                                               3/3       Running            0          33m
openshift-monitoring                     pod/alertmanager-main-2                                               3/3       Running            0          33m
openshift-monitoring                     pod/cluster-monitoring-operator-6ff87b5565-fv95s                      1/1       Running            0          42m
openshift-monitoring                     pod/grafana-57f595895d-xqrr2                                          2/2       Running            0          37m
openshift-monitoring                     pod/kube-state-metrics-dcf7dc56d-7tssr                                3/3       Running            0          29m
openshift-monitoring                     pod/node-exporter-h9mcs                                               2/2       Running            0          30m
openshift-monitoring                     pod/node-exporter-sw74x                                               2/2       Running            0          30m
openshift-monitoring                     pod/prometheus-k8s-0                                                  6/6       Running            1          36m
openshift-monitoring                     pod/prometheus-k8s-1                                                  6/6       Running            1          35m
openshift-monitoring                     pod/prometheus-operator-6dc664b84-s6pp6                               1/1       Running            0          34m
openshift-monitoring                     pod/telemeter-client-7cb686cb6-wc2lz                                  1/2       CrashLoopBackOff   10         28m
openshift-operator-lifecycle-manager     pod/catalog-operator-68c769965b-tszqn                                 1/1       Running            0          47m
openshift-operator-lifecycle-manager     pod/olm-operator-844ff6b98-777lc                                      1/1       Running            0          47m
openshift-operator-lifecycle-manager     pod/package-server-7989d5d6d6-hjzf7                                   1/1       Running            1          47m
openshift-sdn                            pod/ovs-2ztqk                                                         1/1       Running            0          42m
openshift-sdn                            pod/ovs-g5k2m                                                         1/1       Running            0          47m
openshift-sdn                            pod/sdn-47lv2                                                         1/1       Running            0          47m
openshift-sdn                            pod/sdn-controller-rgzkn                                              1/1       Running            0          47m
openshift-sdn                            pod/sdn-mxc95                                                         1/1       Running            0          42m
openshift-service-cert-signer            pod/apiservice-cabundle-injector-7b8667df68-fw2pw                     1/1       Running            0          44m
openshift-service-cert-signer            pod/configmap-cabundle-injector-747dffcb95-6dbfv                      1/1       Running            0          44m
openshift-service-cert-signer            pod/service-serving-cert-signer-858dff84bd-68wfl                      1/1       Running            0          45m
tectonic-system                          pod/kube-addon-operator-654588755f-ch8hd                              1/1       Running            0          44m
```

```console
$  make test-extended SUITE=core FOCUS="Secrets should be consumable from pods in volume with defaultMode set"  TEST_EXTENDED_ARGS="-provider=aws -gce-zone=us-east-1"
test/extended/core.sh

[WARNING] REMINDER, EXTENDED TESTS NO LONGER START A CLUSTER.
[WARNING] THE CLUSTER REFERENCED BY THE 'KUBECONFIG' ENV VAR IS USED.

[INFO] Running tests against existing cluster...
[INFO] Running parallel tests N=<default> with focus Secrets should be consumable from pods in volume with defaultMode set
I1108 14:56:32.104962   22274 test.go:85] Extended test version v4.0.0-alpha.0+de9525e-394-dirty
Running Suite: Extended
=======================
Random Seed: 1541717796 - Will randomize all specs
Will run 1250 specs

Running in parallel across 5 nodes

Nov  8 14:56:37.422: INFO: >>> kubeConfig: /home/adahiya/go/src/github.com/openshift/installer/dev/auth/kubeconfig
Nov  8 14:56:37.424: INFO: Waiting up to 30m0s for all (but 0) nodes to be schedulable
Nov  8 14:56:37.455: INFO: Waiting up to 10m0s for all pods (need at least 0) in namespace 'kube-system' to be running and ready
Nov  8 14:56:37.521: INFO: 3 / 3 pods in namespace 'kube-system' are running and ready (0 seconds elapsed)
Nov  8 14:56:37.521: INFO: expected 1 pod replicas in namespace 'kube-system', 1 are Running and Ready.
Nov  8 14:56:37.523: INFO: Waiting for pods to enter Success, but no pods in "kube-system" match label map[name:e2e-image-puller]
Nov  8 14:56:37.523: INFO: Dumping network health container logs from all nodes...
Nov  8 14:56:37.528: INFO: e2e test version: v1.11.0+d4cacc0
Nov  8 14:56:37.529: INFO: kube-apiserver version: v1.11.0+d4cacc0

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Nov  8 14:56:37.828: INFO: Running AfterSuite actions on all node

SS
------------------------------
Nov  8 14:56:37.829: INFO: Running AfterSuite actions on all node


Nov  8 14:56:37.829: INFO: Running AfterSuite actions on all node


[sig-storage] Secrets
  should be consumable from pods in volume with defaultMode set [NodeConformance] [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:684

[BeforeEach] [Top Level]
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/util/test.go:51
[BeforeEach] [sig-storage] Secrets
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:141
STEP: Creating a kubernetes client
Nov  8 14:56:37.796: INFO: >>> kubeConfig: /home/adahiya/go/src/github.com/openshift/installer/dev/auth/kubeconfig
STEP: Building a namespace api object
Nov  8 14:56:38.530: INFO: About to run a Kube e2e test, ensuring namespace is privileged
Nov  8 14:56:38.660: INFO: No PodSecurityPolicies found; assuming PodSecurityPolicy is disabled.
STEP: Waiting for a default service account to be provisioned in namespace
[It] should be consumable from pods in volume with defaultMode set [NodeConformance] [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:684
STEP: Creating secret with name secret-test-8cd635fd-e3a9-11e8-a375-8c1645754cdc
STEP: Creating a pod to test consume secrets
Nov  8 14:56:38.713: INFO: Waiting up to 5m0s for pod "pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc" in namespace "e2e-tests-secrets-jwtzf" to be "success or failure"
Nov  8 14:56:38.723: INFO: Pod "pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc": Phase="Pending", Reason="", readiness=false. Elapsed: 10.823619ms
Nov  8 14:56:40.733: INFO: Pod "pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc": Phase="Pending", Reason="", readiness=false. Elapsed: 2.020692639s
Nov  8 14:56:42.744: INFO: Pod "pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc": Phase="Succeeded", Reason="", readiness=false. Elapsed: 4.030899929s
STEP: Saw pod success
Nov  8 14:56:42.744: INFO: Pod "pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc" satisfied condition "success or failure"
Nov  8 14:56:42.748: INFO: Trying to get logs from node adahiya-0-worker-0-22h5z pod pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc container secret-volume-test: <nil>
STEP: delete the pod
Nov  8 14:56:42.780: INFO: Waiting for pod pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc to disappear
Nov  8 14:56:42.783: INFO: Pod pod-secrets-8cd74ba9-e3a9-11e8-a375-8c1645754cdc no longer exists
[AfterEach] [sig-storage] Secrets
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:142
Nov  8 14:56:42.783: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
STEP: Destroying namespace "e2e-tests-secrets-jwtzf" for this suite.
Nov  8 14:56:48.816: INFO: Waiting up to 30s for server preferred namespaced resources to be successfully discovered
Nov  8 14:56:49.531: INFO: namespace: e2e-tests-secrets-jwtzf, resource: bindings, ignored listing per whitelist
Nov  8 14:56:51.010: INFO: namespace e2e-tests-secrets-jwtzf deletion completed in 8.222953968s


• [SLOW TEST:13.214 seconds]
[sig-storage] Secrets
/home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/common/secrets_volume.go:33
  should be consumable from pods in volume with defaultMode set [NodeConformance] [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]
  /home/adahiya/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:684
------------------------------
Nov  8 14:56:51.012: INFO: Running AfterSuite actions on all node


Nov  8 14:56:37.829: INFO: Running AfterSuite actions on all node
Nov  8 14:56:51.057: INFO: Running AfterSuite actions on node 1


Ran 1 of 1250 Specs in 13.635 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 1249 Skipped

Ginkgo ran 1 suite in 14.659927285s
Test Suite Passed

[INFO] Running serial tests with focus Secrets should be consumable from pods in volume with defaultMode set
I1108 14:56:51.423741   25221 test.go:85] Extended test version v4.0.0-alpha.0+de9525e-394-dirty
[WARNING] No tests were selected
I1108 14:56:56.307328   27863 test.go:85] Extended test version v4.0.0-alpha.0+de9525e-394-dirty
[WARNING] No tests were selected
[INFO] test/extended/core.sh exited with code 0 after 00h 00m 29s
```

abhinavdahiya commented:
rebased around #510

abhinavdahiya commented on pkg/types/installconfig.go:
```diff
@@ -101,20 +102,21 @@ func (p *Platform) Name() string {
+	// been called ClusterCIDR. If no ClusterNetworks are specified,
+	// we will fall back to the PodCIDR
+	// TODO(cdc) remove this.
+	PodCIDR *ipnet.IPNet `json:"podCIDR,omitempty"`
```
can we open an issue tracking this. JIRA or github either one is fine.

squeed replied:
SDN-235

abhinavdahiya commented:
from ci one of the masters failed to start
```
[core@ip-10-0-29-236 ~]$ sudo crictl ps -a
CONTAINER ID        IMAGE                                                                                                                          CREATED              STATE               NAME                       ATTEMPT
6d0573b005913       ebe29fb54987cd0e79083f08516b143da46fe6d833c6e45d1402147cca6cc40c                                                               12 seconds ago       Exited              openvswitch                8
8ac7d18eae6bf       ebe29fb54987cd0e79083f08516b143da46fe6d833c6e45d1402147cca6cc40c                                                               23 seconds ago       Exited              sdn                        8
57e4647553a60       53503b1b1921a5ec442b4973f8c61954c4b7c01df7c13a01f25b54fe3798fd2a                                                               About a minute ago   Exited              machine-config-daemon      6
728366f0e4a51       registry.svc.ci.openshift.org/ci-op-32yx8i0s/stable@sha256:c6b124d8c65860282900498c275f0c2fca99a83a811e9fbafcfc6b9a5cba76bf    14 minutes ago       Running             machine-config-server      0
c836ca608b38b       registry.svc.ci.openshift.org/ci-op-32yx8i0s/stable@sha256:0143df4b8b0f85c919ea74158c20721a6eaea2126f2e20b5f398cea683de3787    16 minutes ago       Running             sdn-controller             0
013b6ee9a9537       registry.svc.ci.openshift.org/ci-op-32yx8i0s/stable@sha256:48f97ab5e303cb743421194f1c6e73fe58ab4f3cb7728a61cfa48ffdfdd8caa0    17 minutes ago       Running             cluster-network-operator   0
5fee81cfe842f       registry.svc.ci.openshift.org/ci-op-32yx8i0s/release@sha256:b75a76a75a63fee7b68b2d5df9416192abc2516191e363e6d6b4f7407bf60d3a   17 minutes ago       Running             cluster-version-operator   0
b81c0c1f4f325       quay.io/coreos/etcd@sha256:688e6c102955fe927c34db97e6352d0e0962554735b2db5f2f66f3f94cfe8fd1                                    20 minutes ago       Running             etcd-member                0
40a5d41be3eea       quay.io/coreos/kube-client-agent@sha256:8564ab65bcb1064006d2fc9c6e32a5ca3f4326cdd2da9a2efc4fb7cc0e0b6041                       20 minutes ago       Exited              certs                      0
4421c8ce7663f       registry.svc.ci.openshift.org/openshift/origin-v4.0@sha256:07d848b97b3336baf94cc578b7304c29d649cd58aaf4ff71c475b19183a1911b    20 minutes ago       Exited              discovery                  0
[core@ip-10-0-29-236 ~]$ sudo crictl logs 8ac7d18eae6bf
net.ipv4.ip_forward = 1
2018/11/09 00:09:02 socat[4746] E connect(5, AF=1 "/var/run/openshift-sdn/cni-server.sock", 40): No such file or directory
cp: cannot create regular file '/host/opt/cni/bin/host-local': Permission denied
cp: cannot create regular file '/host/opt/cni/bin/loopback': Permission denied
cp: cannot create regular file '/host/opt/cni/bin/openshift-sdn': Permission denied

[core@ip-10-0-29-236 ~]$ sudo crictl logs 6d0573b005913
/etc/openvswitch/conf.db does not exist ... (warning).
Creating empty database /etc/openvswitch/conf.db ovsdb-tool: I/O error: /etc/openvswitch/conf.db: failed to lock lockfile (Resource temporarily unavailable)
[FAILED]

```

abhinavdahiya commented:
Also i'm seeing a lot more restarts with errors:
```console
[core@adahiya-0-master-0 ~]$ sudo crictl logs de51ec38211cd
I1109 00:41:46.139173    6138 start.go:46] Version: 3.11.0-222-g393cc444-dirty
I1109 00:41:46.145260    6138 run.go:22] Running captured: chroot /rootfs rpm-ostree status --json
I1109 00:41:46.177259    6138 daemon.go:93] Booted osImageURL: <not pivoted> (47.29)
F1109 00:41:46.179287    6138 start.go:109] failed to initialize daemon: Failed to set initial annotations: Unable to update node "adahiya-0-master-0": Get https://10.3.0.1:443/api/v1/nodes/adahiya-0-master-0: dial tcp 10.3.0.1:443: connect: connection refused
[core@adahiya-0-master-0 ~]$ sudo crictl logs 40f46b9ce0bf1
W1109 00:41:58.517170       1 authentication.go:237] Unable to get configmap/extension-apiserver-authentication in kube-system.  Usually fixed by 'kubectl create rolebinding -n kube-system ROLE_NAME --role=extension-apiserver-authentication-reader --serviceaccount=YOUR_NS:YOUR_SA'
F1109 00:41:58.517636       1 cmd.go:78] Get https://10.3.0.1:443/api/v1/namespaces/kube-system/configmaps/extension-apiserver-authentication: dial tcp 10.3.0.1:443: connect: connection refused
[core@adahiya-0-master-0 ~]$
```

squeed commented:
Weird! Looks like a SELinux issue. Are we installing policies via cluster components (I thought they were just installed via ostree!?)

squeed commented:
Yes, as expected, it's a SELinux policy deny:

```
selinux is blocking on nodes: Nov 09 12:46:42 ip-10-0-159-158 kernel: type=1400 audit(1541767602.429:6): avc:  denied  { create } for  pid=1682 comm="cp" name="host-local" scontext=system_u:system_r:container_t:s0:c643,c685 tcontext=system_u:object_r:container_var_lib_t:s0 tclass=file
```

Odd that it only happens on worker nodes, and if I manually delete the pods, they start correctly. @crawford / @lucab, do you know of anything that is changing the SELinux policy after the kubelet comes up?

The workaround is just to set `restartPolicy: Always` on the sdn processes.

squeed commented:
Thanks to @lucab's amazing investigation, it seems that there's some weird relabeling race in crio. We don't see it on the masters, because they are up long enough before the sdn starts. But we're one of the first containers to run on the worker nodes.

The "real" solution is to find the crio bug. The horrible solution, which I'm writing now, is to sleep and retry in the init script.

squeed commented:
Just checked, this CI run has the selinux fix. Now we wait.

knobunc commented:
I'd like it to have approved and lgtmed so that if this CI passes we would be ready to go... but I'm not sure if @crawford @wking @smarterclayton @abhinavdahiya have other concerns.

squeed commented:
My fix doesn't really work - it turns out, if the creation is blocked by SELinux, then it will never succeed. We have to exit and start over. The problem is that we risk getting stuck in CrashLoopBackoff.

I'm basically out of ideas. I can tweak the timings, but that seems really sub-optimal.

squeed commented:
/hold

squeed commented:
Putting on hold in case this happens to pass CI, as I was playing with the spun-up CI cluster.

sjenning commented:
weird, everything is Running for me with this patch
```
NAMESPACE                                NAME                                                              READY     STATUS              RESTARTS   AGE
kube-system                              etcd-member-dev-master-0                                          1/1       Running             0          13m
kube-system                              kube-scheduler-ww55g                                              1/1       Running             0          13m
kube-system                              metrics-server-5767bfc576-sl8xh                                   2/2       Running             0          10m
openshift-apiserver                      apiserver-gb9nb                                                   1/1       Running             0          10m
openshift-cluster-api                    clusterapi-manager-controllers-569db585dd-pn6jk                   3/3       Running             0          11m
openshift-cluster-api                    machine-api-operator-64c5f846fb-9j6jr                             1/1       Running             0          13m
openshift-cluster-dns-operator           cluster-dns-operator-56dccb87fd-c9jss                             1/1       Running             0          13m
openshift-cluster-dns                    dns-default-drsmv                                                 1/1       Running             0          12m
openshift-cluster-dns                    dns-default-fdtwm                                                 1/1       Running             0          10m
openshift-cluster-network-operator	 cluster-network-operator-fcvgx                                    1/1       Running             0          13m
openshift-cluster-node-tuning-operator   cluster-node-tuning-operator-547fd78cb7-hsgxd                     1/1       Running             0          9m
openshift-cluster-samples-operator	 cluster-samples-operator-86bb5fc8b-c976g                          1/1       Running             0          9m
openshift-cluster-version                cluster-version-operator-v8ptl                                    1/1       Running             0          13m
openshift-console                        console-operator-7986944548-nj6hd                                 1/1       Running             0          9m
openshift-console                        openshift-console-54f485594d-jq46b                                1/1       Running             0          9m
openshift-controller-manager             controller-manager-pb7d9                                          1/1       Running             0          11m
openshift-core-operators                 openshift-cluster-kube-apiserver-operator-669998d98b-wq69r        1/1       Running             0          13m
openshift-core-operators                 openshift-cluster-kube-controller-manager-operator-5747665xgllk   1/1       Running             0          13m
openshift-core-operators                 openshift-cluster-kube-scheduler-operator-7c9f96ccdb-jh9bh        1/1       Running             0          13m
openshift-core-operators                 openshift-cluster-openshift-apiserver-operator-74649546d7-jhj47   1/1       Running             0          13m
openshift-core-operators                 openshift-cluster-openshift-controller-manager-operator-58lb5sm   1/1       Running             0          13m
openshift-core-operators                 openshift-service-cert-signer-operator-5d4986cf9b-bnpw6           1/1       Running             0          13m
openshift-csi-operator                   csi-operator-66dfcb498b-j68xl                                     1/1       Running             0          9m
openshift-image-registry                 cluster-image-registry-operator-5fd797cc66-n5djp                  1/1       Running             0          9m
openshift-ingress-operator               ingress-operator-5b477c8577-579f2                                 1/1       Running             0          9m
openshift-ingress                        router-default-fgbbr                                              1/1       Running             0          9m
openshift-kube-apiserver                 installer-1-dev-master-0                                          0/1       Completed           0          11m
openshift-kube-apiserver                 openshift-kube-apiserver-dev-master-0                             1/1       Running             0          11m
openshift-kube-controller-manager        installer-1-dev-master-0                                          0/1       Completed           0          11m
openshift-kube-controller-manager        openshift-kube-controller-manager-dev-master-0                    1/1       Running             0          11m
openshift-kube-scheduler                 installer-1-dev-master-0                                          0/1       Completed           0          11m
openshift-kube-scheduler                 openshift-kube-scheduler-dev-master-0                             1/1       Running             0          11m
openshift-machine-config-operator        machine-config-controller-69b6f4858f-7v958                        1/1       Running             0          11m
openshift-machine-config-operator        machine-config-daemon-m6clr                                       1/1       Running             1          10m
openshift-machine-config-operator        machine-config-daemon-wh7hs                                       1/1       Running             2          11m
openshift-machine-config-operator        machine-config-operator-7b97f94d8f-d9dnc                          1/1       Running             0          13m
openshift-machine-config-operator        machine-config-server-6l4fn                                       1/1       Running             0          11m
openshift-monitoring                     alertmanager-main-0                                               3/3       Running             0          5m
openshift-monitoring                     alertmanager-main-1                                               3/3       Running             0          5m
openshift-monitoring                     alertmanager-main-2                                               3/3       Running             0          4m
openshift-monitoring                     cluster-monitoring-operator-5675f6fdb4-5pxkh                      1/1       Running             0          9m
openshift-monitoring                     grafana-57f595895d-dsz4b                                          2/2       Running             0          8m
openshift-monitoring                     kube-state-metrics-dcf7dc56d-8jscb                                3/3       Running             0          4m
openshift-monitoring                     node-exporter-6knnn                                               2/2       Running             0          4m
openshift-monitoring                     node-exporter-bbxzw                                               2/2       Running             0          4m
openshift-monitoring                     prometheus-k8s-0                                                  6/6       Running             1          7m
openshift-monitoring                     prometheus-k8s-1                                                  6/6       Running             1          7m
openshift-monitoring                     prometheus-operator-77f5dc6cc6-q4prr                              1/1       Running             0          5m
openshift-monitoring                     telemeter-client-747b776f55-8zhhp                                 0/3       ContainerCreating   0          4m
openshift-operator-lifecycle-manager     catalog-operator-d5c45d49d-gkh6q                                  1/1       Running             0          13m
openshift-operator-lifecycle-manager     olm-operator-66c4c5b588-8lpg6                                     1/1       Running             0          13m
openshift-operator-lifecycle-manager     package-server-67df777474-zhzsr                                   1/1       Running             0          13m
openshift-sdn                            ovs-vlpdx                                                         1/1       Running             0          12m
openshift-sdn                            ovs-xh5nq                                                         1/1       Running             0          10m
openshift-sdn                            sdn-67rrr                                                         1/1       Running             0          10m
openshift-sdn                            sdn-7tqtn                                                         1/1       Running             0          12m
openshift-sdn                            sdn-controller-qmq2n                                              1/1       Running             0          12m
openshift-service-cert-signer            apiservice-cabundle-injector-57c99484dc-kvrwb                     1/1       Running             0          12m
openshift-service-cert-signer            configmap-cabundle-injector-7f87685598-pjbks                      1/1       Running             0          11m
openshift-service-cert-signer            service-serving-cert-signer-c4db67789-97qnz                       1/1       Running             0          12m
tectonic-system                          kube-addon-operator-654588755f-rkxd8                              1/1       Running             0          13m
```

squeed commented:
> @squeed the old network operator also mounted host paths, can we skip the selinux by moving to a permissive location?

We write to the same location (it's hard-coded in to crio). I wonder why that one works. Good point!

I'll look in to this on Monday.

knobunc commented:
@lucab do you know why there is any relabeling going on?  It sounds wrong for the pod to be relabeling the host filesystem.  Or am I misunderstanding what the problem is.  Thanks!

knobunc commented:
@squeed what if we had an init container run the loop to test?

squeed commented:
Getting closer. In the bad runs, the container processes are started with context `container_t`. In the good runs, they start with `spc_t`. Not sure why this would change between executions. Thoughts, @mrunalp?

squeed commented:
@lucab suggests removing `runAsUser: 0`, which I'm trying now.

squeed commented:
Just waiting for the network operator changes to be picked up by the release, then we can retest.

abhinavdahiya commented:
/retest

squeed commented:
All green. Just needs a lgtm.

abhinavdahiya commented:
The PR is green and changes look fine. We can iterate on this. :yay:

/lgtm

rhatdan commented:
Might have discovered a bug though.  If RunAsUser was causing `privileged` to be ingored?
